kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/ProducerStateManager.java

```diff
@@ -0,0 +1,1048 @@
+            buffer.putLong(producerId);
+            buffer.putLong(firstOffset);
+            buffer.putLong(lastOffset);
+            buffer.putLong(lastStableOffset);
+            buffer.flip();
+            return buffer;
+        }
+    }
+
+    /**
+     * This class is used to validate the records appended by a given producer before they are written to the log.
+     * It is initialized with the producer's state after the last successful append, and transitively validates the
+     * sequence numbers and epochs of each new record. Additionally, this class accumulates transaction metadata
+     * as the incoming records are validated.
+     */
+    public static class ProducerAppendInfo {
+
+        private final String topicPartition;
+
+        // The id of the producer appending to the log
+        private final Long producerId;
+
+        // The current entry associated with the producer id which contains metadata for a fixed number of
+        // the most recent appends made by the producer. Validation of the first incoming append will
+        // be made against the latest append in the current entry. New appends will replace older appends
+        // in the current entry so that the space overhead is constant.
+        private final ProducerStateEntry currentEntry;
+
+        // Indicates the origin of to append which implies the extent of validation.
+        // For example, offset commits, which originate from the group coordinator,
+        // do not have sequence numbers and therefore only producer epoch validation is done.
+        // Appends which come through replication are not validated (we assume the validation has already been done)
+        // and appends from clients require full validation.
+        private final AppendOrigin origin;
+        private final List<TxnMetadata> transactions = Lists.newArrayList();
+        private ProducerStateEntry updatedEntry;
+
+        public ProducerAppendInfo(String topicPartition, Long producerId,
+                                  ProducerStateEntry currentEntry, AppendOrigin origin) {
+            this.topicPartition = topicPartition;
+            this.producerId = producerId;
+            this.currentEntry = currentEntry;
+            this.origin = origin;
+
+            resetUpdatedEntry();
+        }
+
+        private void maybeValidateDataBatch(Short producerEpoch, Integer firstSeq) {
+            checkProducerEpoch(producerEpoch);
+            if (origin.equals(AppendOrigin.Client)) {
+                checkSequence(producerEpoch, firstSeq);
+            }
+        }
+
+        private void checkProducerEpoch(Short producerEpoch) {
+            if (producerEpoch < updatedEntry.producerEpoch) {
+                String message = String.format("Producer's epoch in %s is %s, which is smaller than the last seen "
+                        + "epoch %s", topicPartition, producerEpoch, currentEntry.producerEpoch);
+                throw new IllegalArgumentException(message);
```

Comment:
Though current Kafka 2.0 dependency doesn't have the `InvalidProducerEpochException`, but we can implement it. Then in `KafkaRequestHandler#handlePartitionRecords`, we can catch this exception and return an `INVALID_PRODUCER_EPOCH` error. BTW, I think we should call `Errors.forException` in the catch block instead of just calling `completeExceptionally`.